sklearn_numba_dpex/kmeans/kernels.py

```diff
@@ -0,0 +1,745 @@
+            )
+        )
+
+    def sum_reduction(v):
+        for sum_fn, w in _steps_data:
+            sum_fn(v, w)
+            v = w
+        return v
+
+    return sum_reduction
+
+
+@lru_cache
+def make_sum_reduction_2dim_kernel(n_samples, n_features, work_group_size, dtype):
+    global_size = math.ceil(n_samples / work_group_size) * work_group_size
+    zero = dtype(0.0)
+
+    @dpex.kernel
+    def sum_reduction_kernel(v, w):
+        sample_idx = dpex.get_global_id(0)
+        if sample_idx >= n_samples:
+            return
+        tmp = zero
+        for feature_idx in range(n_features):
+            tmp += v[feature_idx, sample_idx]
+        w[sample_idx] = tmp
+
+    return sum_reduction_kernel[global_size, work_group_size]
+
+
+@lru_cache
+def make_sum_reduction_3dim_kernel(dim0, dim1, dim2, work_group_size, dtype):
+    nb_groups_for_dim2 = math.ceil(dim2 / work_group_size)
+    nb_threads_for_dim2 = nb_groups_for_dim2 * work_group_size
+    global_size = nb_threads_for_dim2 * dim1
+    zero = dtype(0.0)
+
+    @dpex.kernel
+    def sum_reduction_kernel(v, w):
+        group = dpex.get_group_id(0)
+        thread = dpex.get_local_id(0)
+        d1 = group // nb_groups_for_dim2
+        d2 = ((group % nb_groups_for_dim2) * work_group_size) + thread
+        if d2 >= dim2:
+            return
+        tmp = zero
+        for d in range(dim0):
+            tmp += v[d, d1, d2]
+        w[d1, d2] = tmp
+
+    return sum_reduction_kernel[global_size, work_group_size]
+
+
+@lru_cache
+def make_fused_fixed_window_kernel(
+    n_samples,
+    n_features,
+    n_clusters,
+    preferred_work_group_size_multiple,
+    global_mem_cache_size,
+    centroids_window_width_multiplier,
+    centroids_window_height,
+    centroids_private_copies_max_cache_occupancy,
+    work_group_size,
+    dtype,
+):
+    window_nb_centroids = (
+        preferred_work_group_size_multiple * centroids_window_width_multiplier
+    )
+    nb_of_window_features_per_work_group = work_group_size // window_nb_centroids
+
+    centroids_window_height_ratio_multiplier = (
+        centroids_window_height // nb_of_window_features_per_work_group
+    )
+
+    window_nb_features = centroids_window_height
+
+    nb_windows_per_feature = math.ceil(n_clusters / window_nb_centroids)
+    global_size = (math.ceil(n_samples / work_group_size)) * (work_group_size)
+    nb_windows_per_centroid = math.ceil(n_features / window_nb_features)
+    centroids_window_shape = (window_nb_features, (window_nb_centroids + 1))
+
+    inf = dtype(math.inf)
+    zero = dtype(0.0)
+    one = dtype(1.0)
+
+    nb_cluster_items = n_clusters * (n_features + 1)
+    nb_cluster_bytes = 4 * nb_cluster_items
+    # TODO: control that this value is not higher than the number of sub-groups of size
+    # preferred_work_group_size_multiple that can effectively run concurrently. We
+    # should fetch this information and apply it here.
+    nb_centroids_private_copies = int(
```

Comment:
I think you should replace all `nb_` occurrences by `n_` occurrences to be consistent with scikit-learn's naming style.